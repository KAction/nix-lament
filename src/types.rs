use anyhow;
use serde::Serialize;
use tree_sitter as ts;
use tree_sitter_nix as nix;

/// Kind of the lamentations. This is necessary to run only requested subset of supported
/// lamentations.
#[derive(Debug, Serialize)]
pub enum Kind {
    D001,  // Both "pname" and "name" in a call to "mkDerivation"
    PY001, // Explicit "pythonImportsCheckHook" in nativeBuildInputs
    RFC0169, // Naming of the feature parameters
}

// Struct that describes location of the problematic code.
#[derive(Debug, Serialize)]
pub struct Lamentation {
    // By convention, lines are counted from one, columns -- from zero.
    pub line: usize,
    pub column: usize,
    pub kind: Kind,
    pub message: String,
}

pub type Handler = fn(m: &ts::QueryMatch, content: &[u8]) -> Option<Lamentation>;
pub struct PerMatch {
    query: ts::Query,
    handler: Handler,
}

impl PerMatch {
    pub fn new(query: &'static str, handler: Handler) -> anyhow::Result<Self> {
        Ok(Self {
            query: ts::Query::new(nix::language(), query)?,
            handler,
        })
    }

    pub fn lament(&self, tree: &ts::Tree, content: &[u8]) -> Vec<Lamentation> {
        let mut out = vec![];
        let mut cursor = ts::QueryCursor::new();

        for m in cursor.matches(&self.query, tree.root_node(), content) {
            match (self.handler)(&m, content) {
                Some(x) => out.push(x),
                None => (),
            }
        }
        out
    }
}

#[derive(Clone, Copy)]
pub struct Module {
    pub kinds: &'static [Kind],
    pub new: fn() -> anyhow::Result<PerMatch>,
}
