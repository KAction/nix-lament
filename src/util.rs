use crate::types;
use anyhow;
use tree_sitter as ts;
use tree_sitter_nix as nix;

pub fn node_text<'a>(node: &ts::Node, content: &'a [u8]) -> &'a str {
    std::str::from_utf8(&content[node.start_byte()..node.end_byte()]).unwrap()
}

type Handler = fn(m: &ts::QueryMatch, content: &[u8]) -> Option<types::Lamentation>;
pub struct PerMatch {
    query: ts::Query,
    handler: Handler,
}

impl PerMatch {
    pub fn new(query: &'static str, handler: Handler) -> anyhow::Result<Box<dyn types::Lament>> {
        Ok(Box::new(Self {
            query: ts::Query::new(nix::language(), query)?,
            handler,
        }))
    }
}

impl types::Lament for PerMatch {
    fn lament(&self, tree: &ts::Tree, content: &[u8]) -> Vec<types::Lamentation> {
        let mut out = vec![];
        let mut cursor = ts::QueryCursor::new();

        for m in cursor.matches(&self.query, tree.root_node(), content) {
            match (self.handler)(&m, content) {
                Some(x) => out.push(x),
                None => (),
            }
        }
        out
    }
}

#[macro_export]
macro_rules! via_match (
    ($name:ident, $handler:expr) => {
        fn new() -> anyhow::Result<Box<dyn crate::types::Lament>> {
            crate::util::PerMatch::new(include_str!(concat!(stringify!($name), ".scm")), $handler)
        }
        pub static MODULE: crate::types::Module = crate::types::Module {
            new, kinds: &[crate::types::Kind::$name],
        };
    };
);
