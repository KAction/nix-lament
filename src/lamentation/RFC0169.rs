use crate::types;
use crate::util::node_text;
use crate::via_match;
use phf::phf_map;

static REPLACEMENTS: phf::Map<&'static str, &'static str> = phf_map! {
    "alsaSupport" => "with_alsa",
    "enableAlsa" => "with_alsa",
};

via_match!(RFC0169, |m, content| {
    let deprecated = &m.captures[1].node;
    let deprected_name = node_text(deprecated, content);
    let point = deprecated.start_position();
    let new_name = REPLACEMENTS.get(deprected_name).expect(&format!("Replacement for `{}' not found", deprected_name));
    let message = format!("Deprecated feature parameter `{}', use `{}` name instead", deprected_name, new_name);

    Some(types::Lamentation {
        kind: types::Kind::RFC0169,
        line: point.row,
        column: point.column,
        message,
    })
});
