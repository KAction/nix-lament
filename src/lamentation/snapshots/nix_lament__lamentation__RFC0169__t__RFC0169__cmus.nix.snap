---
source: src/lamentation/RFC0169.rs
expression: "w.lament(&tree, &content[..])"
---
- line: 3
  column: 2
  kind: RFC0169
  message: "Deprecated feature parameter `alsaSupport', use `with_alsa` name instead"

